resolvers/dns-resolver/src/dns_client.rs

```diff
@@ -0,0 +1,245 @@
+}
+
+#[derive(Debug, Default)]
+pub struct SrvRecordResp {
+    pub answers: Vec<SrvRecord>,
+    pub additional: Vec<ARecord>,
+}
+
+pub trait DnsClient {
+    fn query_srv(
+        &self,
+        name: &str,
+        log: &Logger,
+    ) -> Result<SrvRecordResp, ResolverError>;
+    fn query_a(
+        &self,
+        name: &str,
+        log: &Logger,
+    ) -> Result<ARecordResp, ResolverError>;
+}
+
+pub struct TrustDnsClient(SyncClient<UdpClientConnection>);
+
+NewtypeFrom! { () pub struct TrustDnsClient(SyncClient<UdpClientConnection>); }
+
+impl DnsClient for TrustDnsClient {
+    fn query_srv(
+        &self,
+        name: &str,
+        log: &Logger,
+    ) -> Result<SrvRecordResp, ResolverError> {
+        let n = Name::from_utf8(name)?;
+        let resp = self.0.query(&n, DNSClass::IN, RecordType::SRV)?;
+        let mut srv_resp = SrvRecordResp::default();
+        debug!(log, "srv_response: {:?}", resp);
+        if resp.response_code() == ResponseCode::NoError {
+            for srv in resp.answers() {
+                if let RData::SRV(ref s_rec) = *srv.rdata() {
+                    srv_resp.answers.push(SrvRecord {
+                        port: s_rec.port(),
+                        priority: s_rec.priority(),
+                        target: s_rec.target().to_string(),
+                        ttl: srv.ttl(),
+                        weight: s_rec.weight(),
+                    });
+                }
+            }
+        }
+        Ok(srv_resp)
+    }
+
+    fn query_a(
+        &self,
+        name: &str,
+        log: &Logger,
+    ) -> Result<ARecordResp, ResolverError> {
+        let n = Name::from_utf8(name)?;
+        let resp = self.0.query(&n, DNSClass::IN, RecordType::A)?;
+        debug!(log, "a resp: {:?}", resp);
+        let mut a_resp = ARecordResp::default();
+        if resp.response_code() == ResponseCode::NoError {
+            for a in resp.answers() {
+                if let RData::A(ref ip) = *a.rdata() {
+                    a_resp.answers.push(ARecord {
+                        name: a.name().to_string(),
+                        ttl: a.ttl(),
+                        address: ip.to_string().parse::<IpAddr>()?,
+                    });
+                }
+            }
+        }
+        Ok(a_resp)
+    }
+}
+
+impl From<ClientError> for ResolverError {
+    fn from(e: ClientError) -> Self {
+        ResolverError::DnsClientError { err: e.to_string() }
+    }
+}
+
+impl From<ProtoError> for ResolverError {
+    fn from(e: ProtoError) -> Self {
+        ResolverError::DnsClientError { err: e.to_string() }
+    }
+}
+
+pub fn init_dns_client(
+    resolver: &str,
+) -> Result<Arc<dyn DnsClient + Send + Sync>, ResolverError> {
+    let server = resolver.to_string().parse()?;
+    match UdpClientConnection::new(server) {
+        Ok(conn) => Ok(Arc::new(TrustDnsClient::from(SyncClient::new(conn)))),
+        Err(e) => Err(ResolverError::DnsClientError { err: e.to_string() }),
```

Comment:
isn't this covered by your `impl<ClientError> for ResolverError` above on line 117?

Reply:
Fixed, thanks.